### DISCUSSION TITLE
try to run opentelemetry cpp - with static libs

### DISCUSSION THREAD

**Discussion by yonatan555** [open-telemetry/opentelemetry-cpp#1617] [Q&A]
I am working on a new project that I want to use open telemetry, I downloaded the whole open source,
I ran cmake for this, and got all compilation results in form of lib.a,
I took all the libs, and all the headers file into my code, and while trying to run the example of metrics with os_stream, I got an error

![image](https://user-images.githubusercontent.com/57701623/190145183-ea460324-e5c9-47ac-bc15-ef7868e5fe51.png)

I inculded the header file into my project, I used -- ENABLE_METRICS_PREVIEW

I compiled it in Linux server , gcc compoler  

Thanks.

**Comment by marcalff** [comment]
Thanks for using opentelemetry-cpp.

There are two implementations of metrics:

- the "old" one, implemented under `sdk/include/opentelemetry/sdk/_metrics` (note the underscore)
- the "current" one, implemented under `sdk/include/opentelemetry/sdk/metrics` (without underscore)

The compiling option `ENABLE_METRICS_PREVIEW` is terribly misnamed.
When enabled, the code actually uses the _old_ metrics implementation (because if was the _new_ thing then, at time of writing).

What you are probably looking for is the current metrics implementation, in which case do not set `WITH_METRICS_PREVIEW` in CMake when building opentelemetry-cpp, and do not set `ENABLE_METRICS_PREVIEW` from your project.

We are definitively aware this area is confusing.

There are no plans to change the meaning of `WITH_METRICS_PREVIEW` / `ENABLE_METRICS_PREVIEW` because this would add even more confusion ... instead, once the current metrics implementation is released, the old _metrics code, as well as `WITH_METRICS_PREVIEW` / `ENABLE_METRICS_PREVIEW`, will be removed.

See related issue #1573

That being said, the old implementation contains the missing code, implemented in

```
MeterProvider::MeterProvider(std::string library_name, std::string library_version) noexcept
    : meter_(new Meter(library_name, library_version))
{}
```

in file `sdk/src/_metrics/meter_provider.cc`,

available in library `opentelemetry_metrics_deprecated`,

To use the old _metrics implementation, make sure to link against `opentelemetry_metrics_deprecated`.

For future bug reports, please paste error messages as plain text instead of screen shots.

**Comment by yonatan555** [comment]
ok ,thanks for the answer, I am a junior , so it is my first time I am working with open-source,and try to understand how to work with any open-source I will meet with 
and I have more questions please if you can:
1)
I really didn't get attention to this configuration, 
now after I changed that, and compiled the project again, using cmake , i didn't get the opentelemetry_metrics_deprecated lib.
just to ensure what I need to change to get the lib above is to enable the old/new version of OTLP ?
that is under the 
root folder -> cmakelist -> option(WITH_METRICS_PREVIEW "Whether to build metrics preview" OFF)
and change OFF/ON ?

2) Can I work with the new version using the default option like above and that should work?

3) I want to get an example to work with this,
am I need to work with /example/metrics_simple ?
I see there are 2 cpp files , and got attention that 
one of them using enable_metrics_privew, 
and one is not 
should I choose one of them to work with accordingly to the version I am using?

thank you a lot , (sorry for my english, hope u got me)

**Comment by marcalff** [comment]
Please check INSTALL.md

https://github.com/open-telemetry/opentelemetry-cpp/blob/main/INSTALL.md

Set compiling options when invoking cmake, do not change the source code itself.

**Comment by yonatan555** [comment]
> 
ok got it, 
first I am working just with static-libs, 
and about the exporter, I forgot to compile it , with option of the exporter thanks :)
Hope it will work for me :)

**Comment by yonatan555** [comment]
while compiling using cmkae, the cmake complain about protobuf,zlib,curl,grpc that it couldn't found these libs

**Comment by lalitb** [comment]
@yonatan555  - You need to install these packages before building opentelemetry-cpp. Please refer to platform-specific documentation for these packages on how to install them. If you are on ubuntu, we have scripts to install them which are used in our CI pipeline.

You can also refer to this docker file - https://github.com/open-telemetry/opentelemetry-cpp/blob/main/docker/Dockerfile.debian.deps which internally uses the above-mentioned scripts to install these dependencies.

Hope it helps.

**Comment by yonatan555** [comment]
ok, I am working on windows, I will try to install these libs, thanks for your help.
